Time varying conversions

tdthompson650 opened the pull request:
# Description
This pull request modifies the conversions so that they can vary with time. Materialized views were used to apply and store the converted readings to improve performance. The graphing functions were modified to retrieve the converted values from the views so that they no longer need to compute the conversion on-the-fly(except for raw reading conversions). The CIK table was modified to contain the start/end timestamp for the conversion, and the start_time was added to the composite primary key.

Partly Addresses #896

## Type of change
- [x ] Note merging this changes the database configuration.
- [ x] This change requires a documentation update

## Checklist
- [ x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x ] I have removed text in ( ) from the issue request
- [ x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations
Although conversions may vary with time, each conversion must be one hour in length and start/end on the hour.

minhtriet359 committed:
updated layout for create a conversion modal, changed scripts from CRLF to LF

tdthompson650 committed:
Created meter_hourly_readings_unit view, added new version of meter_line function that uses view, also all of Wendy's changes

tdthompson650 committed:
Merge branch 'oed-csumb-su25:development' into views

tdthompson650 committed:
Created meter_daily_readings_unit view, altered meter_line_readings_unit_v2 function to use said view.

tdthompson650 committed:
Created new function meter_bar_readings_unit_v2 that uses the meter_daily_readings_unit view

tdthompson650 committed:
Created new function meter_3d_readings_unit_v2 that uses the meter_hourly_readings_unit view

tdthompson650 committed:
Created new function meter_compare_readings_unit_v2 that uses the meter_hourly_readings_unit view

minhtriet359 committed:
changed state in create conversion modal to be compitable with new updates

minhtriet359 committed:
created an overview page for daily patterns

minhtriet359 committed:
changed component folder name from days to daily-pattern

tdthompson650 committed:
Added an index to both meter_hourly_readings_unit and meter_daily_readings_unit, altered cik table to reflect previous groups work with time varying conversion, added 3rd version of meter line function based on previous groups implementaion of time varying conversions for testing purposes

tdthompson650 committed:
Added new meter_hourly view for time varying conversions.

tdthompson650 committed:
Meter hourly time varying conversions are complete, changes were merged into meter_hourly_readings_unit view

LeSci committed:
added function meter_hourly_readings_unit_vB which is hourly_readings_unit  implementing CIK table

evan-carey committed:
Initial UI for the weekly conversion patterns feature

Currently read-only. Uses fake data, as the Weeks API is not yet implemented.

tdthompson650 committed:
modified meter_daily_readings_unit to get it's data from meter_hourly_readings_unit so that it no that it no longer has to apply the conversions it self.

tdthompson650 committed:
Merge branch 'views' of https://github.com/oed-csumb-su25/OED-backend into views

LeSci committed:
update error in hourly_readings_unit and modified meter_hourly_readings_unit_vB

evan-carey committed:
Merge branch 'oed-csumb-su25:development' into evan-carey/weekly-pattern

evan-carey committed:
feat(weeks): add help tooltips to week views; change /weeks page link in navbar to "Patterns - Weekly"

evan-carey committed:
Merge remote-tracking branch 'oed-backend/taliaBranch' into evan-carey/weeks-api

evan-carey committed:
feat(weeks): complete weekly conversion patterns feature

* CRUD operations working with back-end
* client-side validation
* delete confirmation
* Spanish and French translations

evan-carey committed:
Merge remote-tracking branch 'oed-backend/taliaBranch' into evan-carey/weeks-api

evan-carey committed:
fix(weeks): create week API request fails because `id` is included

Omit `id` from the default week values model

evan-carey committed:
revert changes to gitignore and docker-compose

docker-compose should not be ignored and should not have to be altered for our time-varying conversion work

evan-carey committed:
fix(database): weeks table must be created before conversion_segments table

LeSci committed:
update meter_hourly_readings_unit_vB

tdthompson650 committed:
Updated meter_line_readings_unit_v3 function so that it can handle time varying conversions for raw readings.

tdthompson650 committed:
Merge branch 'views' of https://github.com/oed-csumb-su25/OED-backend into views

evan-carey committed:
undo formatting changes to docker-compose

evan-carey committed:
undo formatting change to docker-compose

evan-carey committed:
Address code review comments for Weeks feature

weeksApi
* remove unused selectors
* remove extra newline at eof

CreateWeekModalComponent
* Replace toggleModal with separate show and close functions to match rest of codebase
* use generic `handleStringChange` and `handleNumberChange` function to update model from form.
* remove .trim() from week name validation.
* don't autofocus on first input field. This may cause issues for screen readers

EditWeekModalComponent
* use generic `handleStringChange` and `handleNumberChange` function to update model from form.
* remove .trim() from week name validation.
* remove unnecessary validation checks.
* remove unnecessary comments.

WeeksDetailComponent
* fix alignment of loading spinner

WeekViewComponent
* limit length of note visible in card to 30 characters

evan-carey committed:
Revert docker-compose.yml to match development branch

evan-carey committed:
fix(weeks): WeeksDetailComponent does not need a loading spinner

tdthompson650 committed:
Meter_hourly_readings_unit view is no longer dependent on old hourly view and can now handle multiple conversions per reading that may vary in length and/or overlap the reading. Deleted/comented out some views that are no longer in use.

evan-carey committed:
Merge pull request #1 from oed-csumb-su25/evan-carey/weeks-api

Weeks Feature

tdthompson650 committed:
Deleted old functions, did some reformatting.

tdthompson650 committed:
Merge branch 'oed-csumb-su25:development' into views

tdthompson650 committed:
Updated the meter_raw_readings_unit view, it should not contain all the raw converted readings.

tdthompson650 committed:
Created new view meter_hourly_readings_unit_v3, this view aggregates the converted raw data into hourly readings. Moddified line function to use this view for converted hourly data.

tdthompson650 committed:
Added comments, made sure all comments were on their own line (not trailing code), removed trailing whitespace, reformatting.

minhtriet359 committed:
Merge branch 'development' into development

tdthompson650 committed:
Moved the raw view above the hourly view.

tdthompson650 committed:
removed/commented out unused views

tdthompson650 committed:
Removed an old view, created a new houlry view that uses a CTEand only works under the restriction that only 1 conversion is applied to an hourly reading.

tdthompson650 committed:
Merge remote-tracking branch 'wendys/views' into views

tdthompson650 committed:
moddified group views to get data from new meter views so that they no longer need to apply conversions

huss committed:
Get most tests to work with time varying

- Fixed a long-standing bug where inserting cik into the DB
used forEach with async function but that is not allowed.
Converts to for/of loop so all okay.
- Changes meter refresh model functions to use new views.
- Changed compare meter to make sure range is within
the hour as it did before time varying. It now matches group.
- Changed shrink for meters by day to use new view.
- Placed group views after meter views since they use them.
- Fixed tests to use new views. Some need modification.
to properly handle getting data.
- Misc. formatting.

tdthompson650 committed:
removed old/unused views, added comments, renamed g_unit_id to passed_graphic_unit_id in graphing functions, removed index on old view that was cuasing an error

tdthompson650 committed:
formating

tdthompson650 committed:
Fixed a couple bugs that were causing some test to not pass

tdthompson650 committed:
Migration

tdthompson650 committed:
Modified the migration

huss commented on package.json:
```diff
@@ -45,7 +46,8 @@
 		"webData": "node -e 'require(\"./src/server/data/websiteData.js\").insertWebsiteData()'",
-		"addLogMsg": "node -e 'require(\"./src/server/services/addLogMsg.js\").addLogMsgToDB()'"
+		"addLogMsg": "node -e 'require(\"./src/server/services/addLogMsg.js\").addLogMsgToDB()'",
+		"createCikTest": "node -e 'require(\"./src/server/tmp/uploads/createCikTest.js\").createHourly()'"
```
For testing so should be removed.

tdthompson650 replied:
Line removed.

huss commented on docker-compose.yml:
```diff
@@ -21,8 +21,8 @@ services:
             retries: 3
-        # ports:
-        #    - "5432:5432"
+        ports:
```
For testing so should be commented out.

tdthompson650 replied:
Lines commented out.

huss commented on src/server/sql/cik/create_cik_table.sql:
```diff
@@ -4,9 +4,11 @@
-	slope FLOAT,
-	intercept FLOAT,
-	PRIMARY KEY (source_id, destination_id)
+    source_id INTEGER REFERENCES units(id),
```
Best if put back tab indenting.

tdthompson650 replied:
Fixed when resolving merge conflict.

huss commented on src/server/migrations/2.0.0-3.0.0/sql/cik/alter_cik_table.sql:
```diff
@@ -0,0 +1,11 @@
+--Drop the old primary key (it only includes two columns)
```
For all the migration files:
- Should have the standard OED copyright notice.
- Best if use tab indenting

tdthompson650 replied:
Added OED copyright notice to migration files.

huss commented on src/server/sql/reading/create_function_get_3d_readings.sql:
```diff
@@ -30,171 +25,166 @@ $$ LANGUAGE 'plpgsql';
-    slope FLOAT;
+	-- Holds the range of dates for returned data that fits the actual data.
+	requested_range TSRANGE;
+	-- The slope of the conversion from meter to graphing units
```
slope, intercept are no longer used.

tdthompson650 replied:
Removed unused variables.

huss commented on src/server/sql/reading/create_function_get_3d_readings.sql:
```diff
@@ -30,171 +25,166 @@ $$ LANGUAGE 'plpgsql';
-/* Gets group meters graphing data for 3D graphic by returning points that span the requested
-  length of time over the days requested. 
-*/
 CREATE OR REPLACE FUNCTION group_3d_readings_unit (
```
I'm unclear why the function description comment above this line was removed.

tdthompson650 replied:
Comment added back.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -63,293 +65,243 @@ END;
+-- Index on interval, graphic_unit_id, group_id
+CREATE INDEX if not exists idx_group_daily_readings_unit ON group_daily_readings_unit USING GIST(time_interval, graphic_unit_id, group_id);
+
+--Modified to use meter_hourly_readings_unit in stead of old hourly_readings_unit view.
```
in stead -> instead

tdthompson650 replied:
Spelling corrected.

huss commented on src/server/sql/reading/create_function_get_3d_readings.sql:
```diff
@@ -30,171 +25,166 @@ $$ LANGUAGE 'plpgsql';
-	--Holds the desired meter IDs in order to call meter_3d_readings_unit in the query below. 
+	--Holds the desired meter IDs in order to call meter_3d_readings_unit in the query below.
 	meter_ids INTEGER[];
 BEGIN
```
I'm wondering if this unnest, the sum and the use of the meter 3D function could be avoided if used the group_hourly_readings_unit view to get the values. Maybe the other view PR will be doing this.

huss replied:
Since part of the view PR, I'm moving it there.

huss commented on src/server/sql/reading/create_function_get_compare_readings.sql:
```diff
@@ -57,40 +59,32 @@ DECLARE
-		WHERE curr_tsrange @> hourly.time_interval
-		GROUP BY meters.id, c.slope, c.intercept
+			hourly.meter_id AS meter_id,
+			SUM(hourly.reading_rate) AS reading
```
I think it might be useful to put back the second part of the removed comment on flow unit. This applies in the prev_period case too.

tdthompson650 replied:
comments restored.

tdthompson650 replied:
Please double check to make sure I restored the correct comments.

huss commented on src/server/sql/reading/create_function_get_compare_readings.sql:
```diff
@@ -119,29 +113,68 @@ shift: How far back in time to shift the curr_start and curr_end date/time to ge
+		LEFT JOIN curr_period ON gids.id = curr_period.group_id;
 END;
 $$ LANGUAGE 'plpgsql';
+
```
Extra blank lines.

tdthompson650 replied:
Removed white space.

huss commented on src/server/sql/reading/create_function_get_compare_readings.sql:
```diff
@@ -42,9 +42,11 @@ curr_end: When the current/this time period ends for the compare.
 	times to compare.
  */
+-- New version of the meter_compare_readings function that uses the meter_hourly_readings_unit view.
 CREATE OR REPLACE FUNCTION meter_compare_readings_unit (
```
This logic differs from the group version. I'm wondering if they should be parallel? It also mentions null for missing readings and I wonder what happens in this code.

tdthompson650 replied:
Added "AND hourly.group_id = ANY(group_ids)" to curr_period and prev_period of the group function so that it is now logically equivallent to the meter version. Both functions use left joins so that if a readings doesn't exist for a specified time interval the result with have nulls instead of missing rows.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -9,6 +9,8 @@ Trying to only use case statements led to issues so the following functions
 mix case and if statements.
 */
 
+CREATE EXTENSION IF NOT EXISTS btree_gist;
```
Maybe add (above not below) the comment from the old code:
``-- We need a gist index to support the @> operation.``

tdthompson650 replied:
Comment restored.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -63,293 +65,243 @@ END;
 $$ LANGUAGE 'plpgsql';
 
 /*
```
A big comment was removed. I think parts have merit so it should be updated rather than removed.

tdthompson650 replied:
Comment restored and updated.

tdthompson650 replied:
Please double check that the correct comment was restored.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -63,293 +65,243 @@ END;
+		WHEN u.unit_represent = 'quantity'::unit_represent_type THEN
+			(
+			SUM(
+				-- Reading rate in kw
```
You didn't do this but `` in kw`` should be removed since OED can now do many quantity units.

tdthompson650 replied:
Multiple instances of " in kw" removed.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -63,293 +65,243 @@ END;
+		tsrange(gen.interval_start, gen.interval_start + INTERVAL '1 hour', '()') AS time_interval
+
+	FROM readings r
+	INNER JOIN meters m ON r.meter_id = m.id
```
I'm wondering why the comment was removed? It also applies below. I also see that other joins are done but this does help some to explain the process.

tdthompson650 replied:
Comment restored.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -63,293 +65,243 @@ END;
+)
+SELECT
+	m.id AS meter_id,
+  	SUM(bh.reading_rate * c.slope + c.intercept) AS reading_rate,
```
There are spaces at the start of the line.

tdthompson650 replied:
White space removed.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -63,293 +65,243 @@ END;
+JOIN units  u ON u.id = m.unit_id
+JOIN cik c ON c.source_id = m.unit_id AND tsrange(c.start_time, c.end_time, '()') && bh.time_interval
+GROUP BY m.id, graphic_unit_id, bh.time_interval
+ORDER BY meter_id;
```
I noticed the order by was changed and the old version commented on the clustering. Was this deliberate?

tdthompson650 replied:
I do not recal why I chose the ORDER BY clause. I added gen.interval_start back to the clause since it could affect performance. Also restored comment about clustering.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -63,293 +65,243 @@ END;
+	WHERE tsrange(gen.interval_start, gen.interval_start + INTERVAL '1 day', '()') @> h.time_interval
+	GROUP BY h.meter_id, h.graphic_unit_id, gen.interval_start
+	ORDER BY h.meter_id, graphic_unit_id, gen.interval_start;
+ -- Used by the line/bar/compare functions.
```
Please add blank line separator.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -63,293 +65,243 @@ END;
+-- This index sometimes performs faster(for the bar function) than the above index but is likely not worth the additional overhead.
+-- CREATE INDEX if not exists idx_mdr_meter_graphic ON meter_daily_readings_unit (meter_id, graphic_unit_id);
+
+--Modified to use meter_daily_readings_unit in stead of old daily_readings_unit view.
```
in stead -> instead

While not required, I would put group after meter as done above.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -445,28 +399,55 @@ DECLARE
 		IF (current_point_accuracy = 'raw'::reading_line_accuracy) THEN
 			-- Gets raw meter data to graph.
-			RETURN QUERY
+			RETURN QUERY  --Modified to allow for raw time varying conversions.
```
Please put comments on line above code.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -445,28 +399,55 @@ DECLARE
 					-- the 3600 is needed since EPOCH is in seconds.
-					((r.reading / (extract(EPOCH FROM (r.end_timestamp - r.start_timestamp)) / 3600)) * c.slope + c.intercept) 
+					-- Normalize to rate over reading interval
+					 SUM(
```
Extra space before SUM.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -445,28 +399,55 @@ DECLARE
+							-
+							lower(tsrange(c.start_time, c.end_time, '()') * tsrange(r.start_timestamp, r.end_timestamp, '[]'))
+		  					)) / 3600)
+						* (c.slope * (r.reading / (EXTRACT(EPOCH FROM (r.end_timestamp - r.start_timestamp)) / 3600))+ c.intercept)
```
space before +

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -445,28 +399,55 @@ DECLARE
+					 SUM(
+						--Wrapped in SUM to handle multiple matching cik conversions
+						-- Weight by conversion duration(intersection of reading and conversion time ranges is necessary because the conversion may overlap the reading time range)
+						 (EXTRACT(EPOCH FROM (
```
It would  be nice to have more explicit comments on what this code is doing for the cases.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -605,19 +584,35 @@ BEGIN
+				readings.reading_rate,
+				lower(readings.time_interval) AS start_timestamp,
+				upper(readings.time_interval) AS end_timestamp
+
```
Generally have not had blank line here.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -605,19 +584,35 @@ BEGIN
-		GROUP BY gdm.group_id, readings.start_timestamp, readings.end_timestamp
-		-- This ensures the data is sorted
-		ORDER BY readings.start_timestamp ASC;
+	IF (point_accuracy = 'daily'::reading_line_accuracy) THEN
```
Note on the point_accuracy calculated above (cannot put comment there):
Originally the code looped over the meters so all point accuracies had to be the same. Now it is getting it from the group views. Given this, I don't think they all need to be the same. This is what is done for meters above. I think the code (now or later) should be changed to do this.

tdthompson650 replied:
I'm going to leave this for someone else. I didn't do much with the group functions and I am not certain of how it needs to be modified.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -605,19 +584,35 @@ BEGIN
+			WHERE readings.graphic_unit_id = requested_graphic_unit_id
+			AND tsrange(start_stamp, end_stamp, '[]') @> readings.time_interval
+			-- This ensures the data is sorted
+			ORDER BY readings.time_interval ASC;
```
Here and in hourly, there is no grouping by group_id. I think it might interlace group ids without this if multiple ones are requested at the same time. This may be naturally fixed if address point accuracy per comment above.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -632,9 +627,10 @@ bar_width_days: The number of days to use for the bar width.
 CREATE OR REPLACE FUNCTION meter_bar_readings_unit (
 	meter_ids INTEGER[],
-	graphic_unit_id INTEGER,
+	passed_graphic_unit_id INTEGER, -- This is the graphic unit id, changed from graphic_unit_id to avoid confusion with the graphic unit id in the view.
```
Comment above line.

tdthompson650 replied:
Comment moved.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -679,28 +675,26 @@ BEGIN
 	-- end timestamp by that amount so it stops at the desired end timestamp.
 	real_end_stamp := real_end_stamp - bar_width;
 
+	RAISE NOTICE 'real_start_stamp: %, real_end_stamp: %, num_bars: %',
```
Output should be removed.

tdthompson650 replied:
Raise notice removed.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -729,22 +723,51 @@ DECLARE
+	INNER JOIN unnest(group_ids) gids(id) ON dr.group_id = gids.id;
+
+	real_tsrange := tsrange(date_trunc_up('day', start_stamp), date_trunc('day', end_stamp)) * readings_max_tsrange;
+	-- Get the actual start/end time rounded to the nearest day from the range.
```
I think this comment might be better on the line above since that starts the rounding to the day. However, I just realized this relates to code done in another PR so it might get done there.

tdthompson650 replied:
Not adressing since it is more relevant to another PR.

huss reviewed (changes requested):
Thanks to @tdthompson650 for doing this work. Overall, it works well. I have created some examples of expected results and manual testing indicates they are all correct so far. In the future these and others will be integrated into OED testing for more complete checks. There are some comments in the code to consider. Here are a few that could not easily be put into the code:
- There are a couple of merge conflicts.
- This branch seems behind timeVary and also includes code from the view PR. I'm will to work on this issue but wanted to note it. I have not reviewed the code from the other PR as it is being done there.
- src/server/data/automatedTestingData.js line 1349 inserts groups but does not do the equivalent of refreshGroupsDeepMeters. I manually did it and then the results were fine. This needs to be integrated into the code.

I wanted to ask if you want to work on any of the comments given our conversations. I'm happy for you to address the ones you want. It is also okay to indicate you would like them addressed by someone else. Please just let me/OED know.

huss commented:
@tdthompson650 I wanted to note that I just merged in @vyvyvyThao PR. After you finish any changes, I will take care of all the merge conflicts and verify that merging in that code into this PR does not cause any issues.

tdthompson650 committed:
Merge remote-tracking branch 'upstream/development' into views

tdthompson650 committed:
Revert "Merge remote-tracking branch 'upstream/development' into views"

This reverts commit e29aadf610d12fcd76c899096753bf2098d92ba7, reversing
changes made to 2f6e990ef89b861e62dbb1aca5240d37e7bc49c5.

Undoing merge with upstream/development

tdthompson650 committed:
resolved most PR comments

huss committed:
address some comments/minor

- Tab indenting for sql.
- Minor spacing.
- Minor comment wording.
- Reorder meter hour/daily view to be consistent.

huss committed:
Merge remote-tracking branch 'origin/timeVary' into pr/tdthompson650/1546

This version starts up but has issues. It is mixing cik and cik_vary.
It probably has other issues but wanted to commit at this point.

huss committed:
correct that no rows possible in query

The path algorithm checks if segments exist and sometimes they do not if the
conversion is backward so it is many or none (any).

huss committed:
basic working with cik_vary

cik_vary is currently copied to cik to get this temporary version working.
The creation of the deep group views and the redoCikVary and the reading
views is still done manually.

huss committed:
try to fix merge issues

A number of files were messed up during the merge of timeVary. This
attempts to fix them but testing, etc. is not yet working. Unfortunately,
I could not figure out how to keep the original git attribution for most
lines so they how indicate I did it. I want to note that most were originally
done by @tfrendl and earlier commits show this.

huss committed:
fix sql for and conversion test for cik_vary

- The reading views should now depend on cik_vary.
- The conversion tests now properly create conversion and test
more systematically.

huss committed:
fix cik create & compareTest

- The insertion of cik from cik now does unique ones.
- Test does redoCik so meter seen.

huss committed:
use cikVary not cik

- This fixes a lot of tests & other code.
- Suffix units & 3D tests still failing.

huss committed:
fix suffix units

- It was trying to get slope/intercept from conversion but now in
conversionSegment. The code only works if not time-varying so
only one conversionSegment. This will need to be addressed.
- Formatting.

huss committed:
3D group fixed & all tests passing

- The edits to 3D were lost in the merge. Mostly this was the new version
of 3D group but some other minor changes were also carried over.

huss committed:
fix testData so refresh views

Needed to add deep group refresh and do reading views after that.
Also removed unneeded import.

huss committed:
Update cik

- Remove slope/intercept from cik on client and server.
- Note client still has slope/intercept in conversions because new code not
yet merged. This means creating conversion does not work quite right and
they cannot be deleted.
- A few minor updates to files.

huss committed:
update migrations

huss commented:
I wanted to leave a comment about the current status before I need to look at other OED work. The code is passing all the CI tests and a manual one I tried for one TV case. Here is a basic list of what is left at this point:

- Check over all PR changes again due to the large changes since I last did this.
- Look over all comments and TODO items to decide if they should be addressed now, made into nearer-term action items or left as TODO in the code.
- Work on a few remaining items I found.
- Retest for more cases to be sure it generally works.

@tdthompson650 I made non-trivial changes due to the merge and other updates so please let me know if you have any thoughts.
